<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta name="google-site-verification" content="google481e35697c8720a4"> <meta http-equiv="Permissions-Policy" content="interest-cohort=()"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Tutorial: RNA-seq short variant calling using GATK4 | Xiaofei Carl Zang </title> <meta name="author" content="Xiaofei Carl Zang"> <meta name="description" content="Computational Biology, Transcriptomics, and something. "> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="/assets/img/favicon.ico?3816ea69d189345544a27cde9647370c"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://x-zang.github.io/posts/2021/02/gatk-rna-snp/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Xiaofei</span> Carl <span class="font-weight-bold">Zang</span> </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/software/">software </a> </li> <li class="nav-item "> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Tutorial: RNA-seq short variant calling using GATK4</h1> <p class="post-meta"> Created on February 05, 2021 </p> <p class="post-tags"> <i class="fa-solid fa-calendar fa-sm"></i> 2021   ·   <i class="fa-solid fa-hashtag fa-sm"></i> pipline   <i class="fa-solid fa-hashtag fa-sm"></i> tutorial   <i class="fa-solid fa-hashtag fa-sm"></i> gatk </p> </header> <article class="post-content"> <div id="markdown-content"> <p>GATK is powerful. However, running it may not be as easy. People, especially bioinformatics beginners are often overwhelmed by its powerfulness and complexity.</p> <p>This repo is a tutorial of how to locally running the workflow for RNA-seq short variant calling (SNPs &amp; indels) using GATK4. The original workflow is available at <a href="https://github.com/gatk-workflows" rel="external nofollow noopener" target="_blank">gatk-workflows</a>/<strong><a href="https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels" rel="external nofollow noopener" target="_blank">gatk4-rnaseq-germline-snps-indels</a></strong>, developed by the GATK Team. This repo is made for my personal interest and record to make it easier to run GATK workflow. Root is not required if this tutorial is followed. The tutorial is made based on the <a href="https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels" rel="external nofollow noopener" target="_blank">GATK4 workflow repo</a>, <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531192-RNAseq-short-variant-discovery-SNPs-Indels-" rel="external nofollow noopener" target="_blank">its best practice</a>, and the tutorial on <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035530952?id=12521" rel="external nofollow noopener" target="_blank">how to run GATK workflow</a>.</p> <p>Note: This is for GATK4 and may not be compatible with GATK3.8.</p> <p><strong>Last updated: Jul 2, 2020.</strong></p> <h1 id="install-gatk4">Install GATK4</h1> <p><strong>Download the latest release</strong> from GATK4 repository https://github.com/broadinstitute/gatk/releases. Downloading the pre-compiled binary file is the easiest installation. Requirements for GATK4 can be found form https://github.com/broadinstitute/gatk.</p> <div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># download gatk</span>
<span class="c"># replace the link with latest gatk release from https://github.com/broadinstitute/gatk/releases</span>
wget https://github.com/broadinstitute/gatk/releases/download/4.1.8.0/gatk-4.1.8.0.zip
unzip gatk-4.1.8.0.zip
<span class="nb">cd </span>gatk-4.1.8.0
</code></pre></div></div> <p><strong>Install conda environment for GATK</strong>. On <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035889851--How-to-Install-and-use-Conda-for-GATK4" rel="external nofollow noopener" target="_blank">this page</a>, it is stated that the bioconda GATK installation does not configure the environment correctly. Then, it is necessary to manually create a conda environment and install GATK dependencies.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># create a conda environment</span>
conda <span class="nb">env </span>create <span class="nt">-n</span> gatk <span class="nt">-f</span> gatkcondaenv.yml
<span class="c"># or </span>
conda <span class="nb">env </span>create <span class="nt">--prefix</span> ./path/to/directory/ <span class="nt">-f</span> gatkcondaenv.yml
<span class="c"># activate conda environment</span>
conda activate gatk
<span class="c"># or</span>
conda activate ./path/to/directory/
</code></pre></div></div> <p>You might also want to install <code class="language-plaintext highlighter-rouge">java-jdk</code> if it is not yet installed (<code class="language-plaintext highlighter-rouge">conda install -c cyclus java-jdk</code>).</p> <p>Also, have STAR installed in this environment.</p> <p>This is a convenient way of installing required conda dependencies of GATK. Sometimes it doesn’t work, e.g. conflicts. In case of conda not working, manually install the packages described in this file with <code class="language-plaintext highlighter-rouge">conda</code> and <code class="language-plaintext highlighter-rouge">pip</code>.</p> <h1 id="run-docker-without-root">Run Docker without root</h1> <p>Docker is required to run GATK4 workflow. Root is needed, but you can also run docker daemon without sudo (<a href="https://docs.docker.com/engine/security/rootless/" rel="external nofollow noopener" target="_blank">see docker docs for more details</a>). This section works for Ubuntu. Other OS may have some <a href="https://docs.docker.com/engine/security/rootless/#prerequisites" rel="external nofollow noopener" target="_blank">prerequisites</a>.</p> <div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code>curl <span class="nt">-fsSL</span> https://get.docker.com/rootless | sh

<span class="c"># At the end of the script, some environment variables are required to be set.</span>
<span class="c"># They would be displayed like the below examples, C&amp;P to run them.</span>
<span class="nb">export </span><span class="nv">DOCKER_HOST</span><span class="o">=</span>unix:///run/user/1001/docker.sock
</code></pre></div></div> <p>Replace the number in <code class="language-plaintext highlighter-rouge">/run/user/1001/</code> with your uid, which can be retrieved using <code class="language-plaintext highlighter-rouge">id -u</code>. For example, if <code class="language-plaintext highlighter-rouge">id -u</code> outputs <code class="language-plaintext highlighter-rouge">1005</code>, the command should be <code class="language-plaintext highlighter-rouge">export DOCKER_HOST=unix:///run/user/1005/docker.sock</code>.</p> <div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># start/stop/restart docker without root using the following</span>
systemctl <span class="nt">--user</span> <span class="o">(</span>start|stop|restart<span class="o">)</span> docker
</code></pre></div></div> <h3 id="change-docker-root-directory">Change docker root directory</h3> <p>When running docker, I found it uses <code class="language-plaintext highlighter-rouge">$home/.local/share</code> as the root dir. This directory is in my <code class="language-plaintext highlighter-rouge">$home</code>, which is limited in space. I changed docker root dir following <a href="https://medium.com/@hsadanuwan/how-to-change-docker-default-data-directory-f884dac76c1f" rel="external nofollow noopener" target="_blank">this</a>, otherwise for my case, the disk can be full quickly and the process fails.</p> <p>Run <code class="language-plaintext highlighter-rouge">docker info</code>. The line starting with <code class="language-plaintext highlighter-rouge">Docker Root Dir</code> states the root directory of docker. I intend to change it to another directory in the disk. It can be configured by editing <code class="language-plaintext highlighter-rouge">docker.service</code> file.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># find the docker.service file</span>
locate docker.service
</code></pre></div></div> <p>Edit <code class="language-plaintext highlighter-rouge">docker.service</code> file. Find the <code class="language-plaintext highlighter-rouge">ExecStart</code> line and add <code class="language-plaintext highlighter-rouge">-g /customized/root/dir</code> to the end of this line.</p> <p>Now for me, this line looks like <code class="language-plaintext highlighter-rouge">ExecStart=/home/userid/bin/dockerd-rootless.sh --experimental --storage-driver=overlay2 -g /disk/userid/tools/docker-tmp</code></p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># reload dockerd and restart docker</span>
systemctl <span class="nt">--user</span> daemon-reload
systemctl <span class="nt">--user</span> restart docker
</code></pre></div></div> <p>Run <code class="language-plaintext highlighter-rouge">docker info </code> again. I can see <code class="language-plaintext highlighter-rouge">Docker Root Dir: /disk/userid/tools/docker-tmp</code>. Then all’s set. Some other options can be found <a href="https://github.com/IronicBadger/til/blob/master/docker/change-docker-root.md" rel="external nofollow noopener" target="_blank">here</a></p> <h1 id="download-the-gatk4-workflow">Download the GATK4 workflow</h1> <p>Set up a working directory.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">mkdir </span>gatk-workflows
<span class="nb">cd </span>gatk-workflows
</code></pre></div></div> <p>Download the latest releases of the workflow from <a href="https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels/releases" rel="external nofollow noopener" target="_blank">here</a>.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># replace the following link with the latest release</span>
wget https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels/archive/1.0.0.tar.gz 
<span class="c"># change the .tar.gz file name if needed</span>
<span class="nb">tar</span> <span class="nt">-zxvf</span> 1.0.0.tar.gz
</code></pre></div></div> <h1 id="prepare-input-files">Prepare input files</h1> <p>Set up inputs directory, and put all necessary input files in this directory.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">mkdir </span>inputs
</code></pre></div></div> <p>Download necessary file from Google Cloud Bucket of Broad Institute https://console.cloud.google.com/storage/browser/gcp-public-data–broad-references/ using browser or <a href="https://cloud.google.com/storage/docs/gsutil_install#linux" rel="external nofollow noopener" target="_blank">gsutil</a>. Most of the required references and databases can be found there. Make sure all of the files uses either one of UCSC names or Ensembl names.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># This is an example of how to download with gsutil</span>
gsutil <span class="nt">-m</span> <span class="nb">cp </span>gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.dbsnp138.vcf ./inputs/
</code></pre></div></div> <p>The input file format for GATK is <strong>unmapped bam</strong>. GATK <code class="language-plaintext highlighter-rouge">FastqToSam</code> can be used to convert fastq to unmapped bam.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>gatk FastqToSam <span class="se">\</span>
<span class="nt">-F1</span> reads_R1.fastq <span class="se">\</span>
<span class="nt">-F2</span> reads_R2.fastq <span class="se">\</span>
<span class="nt">-O</span> reads.unmapped.bam <span class="se">\</span>
<span class="nt">--SAMPLE_NAME</span> sample001 <span class="se">\</span>
<span class="nt">--PLATFORM</span> ILLUMINA <span class="se">\</span>
<span class="nt">--READ_GROUP_NAME</span> group01 <span class="se">\</span>
<span class="nt">--SORT_ORDER</span> unsorted 
<span class="c"># --SORT_ORDER {unsorted, queryname, coordinate, duplicate, unknown}</span>
</code></pre></div></div> <p>Index and dictionary files (.fai/.dict/.idx) can be generated with samtools and igvtools</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools faidx Homo_sapiens_assembly38.genome.fasta
samtools dict Homo_sapiens_assembly38.genome.fasta <span class="nt">-o</span> Homo_sapiens_assembly38.genome.dict 
igvtools index file.vcf 
</code></pre></div></div> <p><strong>Then edit the <code class="language-plaintext highlighter-rouge">.json</code> file (in the GATK workflow directory <code class="language-plaintext highlighter-rouge">gatk4-rnaseq-germline-snps-indels</code>) to replace the corresponding file paths with your local file paths.</strong></p> <p>Also replace the GATK path in the <code class="language-plaintext highlighter-rouge">.json</code> file with the directory where GATK4 is installed.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  "##_COMMENT5": "PATHS",
  "#RNAseq.gatk_path_override": "/path/to/gatk4",
</code></pre></div></div> <h4 id="edit-gatk4-rna-best-practiceswdl-file"><strong>Edit <code class="language-plaintext highlighter-rouge">gatk4-rna-best-practices.wdl</code> file</strong></h4> <p>I’m not sure if this step is necessary or correct, but it worked for me (version <code class="language-plaintext highlighter-rouge">gatk4-rnaseq-germline-snps-indels-1.0.0</code>). Read the following instructions, or directly replace the original one with my edited <code class="language-plaintext highlighter-rouge">.wdl</code> (version 1.0.0, see <a href="https://github.com/x-zang/gatk4-rnaseq-germline-snps-indels/blob/master/gatk4-rna-best-practices.wdl" rel="external nofollow noopener" target="_blank">this</a>).</p> <p>Search <code class="language-plaintext highlighter-rouge">BedToIntervalList </code> in the <code class="language-plaintext highlighter-rouge">gatk4-rna-best-practices.wdl</code> file. You can see a block of code like the following.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>        ${gatk_path} \
            BedToIntervalList \
            -I=exome.fixed.bed \
            -O=${output_name} \
            -SD=${ref_dict}
</code></pre></div></div> <p>This section looks more like GATK3.8 commands but not GATK4’s, so I deleted the <code class="language-plaintext highlighter-rouge">=</code> sign behind <code class="language-plaintext highlighter-rouge">-I</code>, <code class="language-plaintext highlighter-rouge">-O</code>, <code class="language-plaintext highlighter-rouge">-SD</code> arguments. Then this block of code looks like:</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>        ${gatk_path} \
            BedToIntervalList \
            -I exome.fixed.bed \
            -O ${output_name} \
            -SD ${ref_dict}
</code></pre></div></div> <p>Do the same to the <code class="language-plaintext highlighter-rouge">IntervalListTools</code> code block.</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>        ${gatk_path} --java-options "-Xms1g" \
            IntervalListTools \
            --SCATTER_COUNT ${scatter_count} \
            --SUBDIVISION_MODE BALANCING_WITHOUT_INTERVAL_SUBDIVISION_WITH_OVERFLOW \
            --UNIQUE true \
            --SORT true \
            --INPUT ${interval_list} \
            --OUTPUT out
</code></pre></div></div> <p>And <code class="language-plaintext highlighter-rouge">MergeVcfs</code> (line 721):</p> <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>        ${gatk_path} --java-options "-Xms2000m"  \
            MergeVcfs \
            --INPUT ${sep=' --INPUT ' input_vcfs} \
            --OUTPUT ${output_vcf_name}
</code></pre></div></div> <h1 id="execute-the-workflow">Execute the workflow</h1> <p>Everything should be all set so far. Download the <a href="https://github.com/broadinstitute/cromwell/releases" rel="external nofollow noopener" target="_blank">cromwell</a> program which will execute the GATK4 workflow. <code class="language-plaintext highlighter-rouge">cd</code> to the <code class="language-plaintext highlighter-rouge">gatk-workflows</code> directory.</p> <div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>wget https://github.com/broadinstitute/cromwell/releases/download/51/cromwell-51.jar

<span class="c"># execute GATK4 workflow</span>
java <span class="nt">-jar</span> cromwell-51.jar run gatk4-rnaseq-germline-snps-indels-1.0.0/gatk4-rna-best-practices.wdl <span class="nt">--inputs</span> gatk4-rnaseq-germline-snps-indels-1.0.0/gatk4-rna-germline-variant-calling.inputs.json
</code></pre></div></div> <h1 id="references">References</h1> <p>https://github.com/gatk-workflows/gatk4-rnaseq-germline-snps-indels</p> <p>https://gatk.broadinstitute.org/hc/en-us/articles/360035531192-RNAseq-short-variant-discovery-SNPs-Indels-</p> <p>https://gatk.broadinstitute.org/hc/en-us/articles/360035530952?id=12521</p> <p>https://docs.docker.com/engine/security/rootless/</p> </div> </article> <div id="giscus_thread" style="max-width: 930px; margin: 0 auto;"> <br> <script>
      let giscusTheme = determineComputedTheme();
      let giscusAttributes = {
        src: 'https://giscus.app/client.js',
        'data-repo': 'x-zang/x-zang.github.io',
        'data-repo-id': '',
        'data-category': 'Comments',
        'data-category-id': '',
        'data-mapping': 'title',
        'data-strict': '1',
        'data-reactions-enabled': '1',
        'data-emit-metadata': '0',
        'data-input-position': 'bottom',
        'data-theme': giscusTheme,
        'data-lang': 'en',
        crossorigin: 'anonymous',
        async: '',
      };

      let giscusScript = document.createElement('script');
      Object.entries(giscusAttributes).forEach(([key, value]) => giscusScript.setAttribute(key, value));
      document.getElementById('giscus_thread').appendChild(giscusScript);
    </script> <noscript>Please enable JavaScript to view the <a href="http://giscus.app/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by giscus.</a> </noscript> </div> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Xiaofei Carl Zang. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-79DTGJE6KQ"></script> <script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
      dataLayer.push(arguments);
    }
    gtag('js', new Date());

    gtag('config', 'G-79DTGJE6KQ');
  </script> <script defer src="/assets/js/google-analytics-setup.js"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>