<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta name="google-site-verification" content="google481e35697c8720a4"> <meta http-equiv="Permissions-Policy" content="interest-cohort=()"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Strandness of RNA-seq and Transcripts Explained | Xiaofei Carl Zang </title> <meta name="author" content="Xiaofei Carl Zang"> <meta name="description" content="Computational Biology, Transcriptomics, and something. "> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="/assets/img/favicon.ico?3816ea69d189345544a27cde9647370c"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://x-zang.github.io/posts/2024/09/check-strandess/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Xiaofei</span> Carl <span class="font-weight-bold">Zang</span> </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/software/">software </a> </li> <li class="nav-item "> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Strandness of RNA-seq and Transcripts Explained</h1> <p class="post-meta"> Created on September 26, 2024, last updated on February 28, 2025 </p> <p class="post-tags"> <i class="fa-solid fa-calendar fa-sm"></i> 2024   ·   <i class="fa-solid fa-hashtag fa-sm"></i> rna-seq   <i class="fa-solid fa-hashtag fa-sm"></i> transcriptome </p> </header> <article class="post-content"> <div id="markdown-content"> <p>Two posts by <a href="https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/" rel="external nofollow noopener" target="_blank">Griffith Lab</a> and by <a href="https://littlebitofdata.com/en/2017/08/strandness_in_rnaseq/" rel="external nofollow noopener" target="_blank">Hong Zheng</a> are already pretty comprehensive in describing the strandness parameters of many tools. Here I make some notes for my own records.</p> <h1 id="strandness-of-rna-seq">Strandness of RNA-seq</h1> <p>Imagine we collected some RNA molecules, performed the sequencing experiments, and obtained a bunch of paired-end RNA-seq reads (respectively R1 and R2 for each end of one pair). Now we want to know whether R1 has the same sequence as the original RNA or R2 has the same sequence as the original RNA (but of course, the reads are shorter and erroneous compared to an RNA molecule). This is called <strong>strandness</strong> of RNA-seq.</p> <p>As of today, the mainstream of NGS (e.g. Illumina) technology is sequence-by-synthesis (SBS). Consequently, R1 and R2 must be from different strands of the RNA molecule, because they are synthesized from different directions! That means we have only 2 possibilities of strandness: (1) R1 is from sense-strand and R2 is form anti-sense-strand; or (2) R1 is from anti-sense strand and R2 is from sense-strand. Here we call the strand with the <em>same</em> sequences of the RNA as <em>sense</em> strand. Likewise, the strand that is reverse-complementary to the RNA is called <em>anti-sense</em> strand.</p> <h2 id="naming-conventions">Naming conventions</h2> <p>Most of the name conventions fall into two types: <code class="language-plaintext highlighter-rouge">rf/fr</code> for <code class="language-plaintext highlighter-rouge">strandness</code> and <code class="language-plaintext highlighter-rouge">first/second</code> for <code class="language-plaintext highlighter-rouge">library type</code>.</p> <p><strong>RF</strong>: The first read is <strong>R</strong>everse sequence of RNA and the second read is <strong>F</strong>orward sequence of RNA.</p> <p><strong>FR</strong>: The first read is <strong>F</strong>orward sequence of RNA and the second read is <strong>R</strong>everse sequence of RNA.</p> <p>The <code class="language-plaintext highlighter-rouge">first/second</code> library types can be somewhat confusing (at least I got confused a few times…). R1 of <code class="language-plaintext highlighter-rouge">first</code> stranded libraries are the same as the <em>anti-sense</em> strand and vice versa. This is because the sequenced substrates in RNA-seq technologies are (for most of the time) cDNAs. The first cDNA strand, which uses the RNA as template, is rev-comp to the RNA sequence. Then cDNA molecules can be synthesized to be double-stranded. Since cDNAs are double-stranded, this further divides technologies by whether:</p> <p><strong>first</strong>: sequencing the first cDNA strand (anti-sense)</p> <p><strong>second</strong>: sequencing the second cDNA strand (sense).</p> <p>Note that the first cDNA strand is reverse-complementary to the RNA molecule.</p> <h2 id="parameters-for-some-bioinformatics-tools">Parameters for some bioinformatics tools</h2> <p>Two tools, <a href="https://rseqc.sourceforge.net/#infer-experiment-py" rel="external nofollow noopener" target="_blank">infer_experiment.py</a> from RSeQC and <a href="https://github.com/signalbash/how_are_we_stranded_here" rel="external nofollow noopener" target="_blank">check_strandness</a> are easy-to-use tools for checking strandness.</p> <p>Assuming all reads are <em>paired</em> and sequenced <em>inwards</em>.</p> <table> <thead> <tr> <th>RNA</th> <th>5’ –&gt; 3’</th> <th>3’ –&gt; 5’</th> <th> </th> </tr> </thead> <tbody> <tr> <td>DNA</td> <td>coding/ sense strand</td> <td>noncoding/ anti-sense</td> <td> </td> </tr> <tr> <td>Reads</td> <td>R1 is a sub-string of RNA, R2 is rev-comp</td> <td>R2 is a sub-string of RNA, R1 is rev-comp</td> <td>Both R1 and R2 may be the same or rev-comp of a sub-string of RNA</td> </tr> <tr> <td>infer_experiment.py</td> <td>1++,1–,2+-,2-+</td> <td>1+-,1-+,2++,2–</td> <td> </td> </tr> <tr> <td>check_strandness</td> <td>FR/fr-secondstrand</td> <td>RF/fr-firststrand</td> <td> </td> </tr> <tr> <td>Kallisto</td> <td><code class="language-plaintext highlighter-rouge">--fr-stranded</code></td> <td><code class="language-plaintext highlighter-rouge">--rf-stranded</code></td> <td> </td> </tr> <tr> <td>Salmon</td> <td><code class="language-plaintext highlighter-rouge">--libType ISF</code></td> <td><code class="language-plaintext highlighter-rouge">--libType ISR</code></td> <td><code class="language-plaintext highlighter-rouge">--libType IU</code></td> </tr> <tr> <td>Scallop</td> <td><code class="language-plaintext highlighter-rouge">--library_type second</code></td> <td><code class="language-plaintext highlighter-rouge">--library_type first</code></td> <td><code class="language-plaintext highlighter-rouge">--library_type unstranded</code></td> </tr> <tr> <td>Stringtie</td> <td><code class="language-plaintext highlighter-rouge">--fr</code></td> <td><code class="language-plaintext highlighter-rouge">--rf</code></td> <td> </td> </tr> </tbody> </table> <h1 id="strand-of-transcripts">Strand of transcripts</h1> <p>Transcripts/RNAs also have “strand” information (e.g. the 7th column <code class="language-plaintext highlighter-rouge">strand</code> in a gtf file). It needs to be clarified that:</p> <ul> <li>strandness of RNA-seq: we are talking about which <strong>strand of RNA/cDNA</strong> the <strong>seq reads are from</strong>, the cRNA strand (i.e. rev-comp to RNA) or the direct-RNA strand (i.e. same sequence as RNA). The second strand of cDNA has the same sequence as the direct RNA, so computationally direct-RNA seq has the same strandness as “second-stranded”</li> <li>strand of transcripts: we are talking about which <strong>strand of the genome</strong> the <strong>gene/transcript is from</strong>. In other words, whether the RNA aligns to the forward sequence of the genome (<code class="language-plaintext highlighter-rouge">+</code> strand, same sequence as genome.fa file) or the RNA aligns to the reverse-complementary sequence of the genome (<code class="language-plaintext highlighter-rouge">-</code> strand, rev-comp to sequence of genome.fa file).</li> </ul> <p>The strand of a transcript can be inferred by using read alignment and strandness of reads, and vice versa.</p> <h2 id="sam-tags-ts-tx">SAM tags <code class="language-plaintext highlighter-rouge">ts</code>, <code class="language-plaintext highlighter-rouge">tx</code> </h2> <p>Some splice-aware aligner outputs <a href="https://samtools.github.io/hts-specs/SAMtags.pdf" rel="external nofollow noopener" target="_blank">SAM tags</a> <code class="language-plaintext highlighter-rouge">ts</code> that indicates <em>which transcript strand the read is from</em>. This flag is usually inferred by checking the canonical intron splice motif of the reads without prior knowledge of the transcript information. Namely, <code class="language-plaintext highlighter-rouge">+</code> ts-tag means the read is from the same strand as the mRNA, and <code class="language-plaintext highlighter-rouge">-</code> ts-tag means the read if from the first cDNA strand (rev-comp to mRNA). Likewise, R1 of a paired and <em>fr-stranded</em> RNA-seq sample is supposed to have all <em>positive</em> ts tag, while R2 of the same sample has all <em>negative</em> ts tag. Reads from an unstranded sample may have roughly half reads assigned positive ts and half assigned negative ts.</p> <p>Some splice-aware aligners (e.g. <a href="https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf" rel="external nofollow noopener" target="_blank">STAR</a> by setting <code class="language-plaintext highlighter-rouge">--outSAMstrandField intronMotif</code>) output SAM tags <code class="language-plaintext highlighter-rouge">xs</code> that indicate <em>strand of the RNA transcript</em>. Namely, this xs tag should be the same as the <code class="language-plaintext highlighter-rouge">strand</code> information in a gtf file of the same transcript. The information of ts and xs can be inferred by examining the read alignments. Obviously, if a read aligns to the positive-strand of the genome and it is from the positive strand of the RNA (positive ts), then the transcript should be from the positive strand of the genome (positive xs, positive strand in gtf). If a read aligns to the negative-strand of the genome and it is from the negative strand of the RNA (negative ts), then the transcript should still be from the positive strand of the genome (positive xs, positive strand in gtf), i.e. the double negation cancels out.</p> <p>In SAM/BAM format, bit <code class="language-plaintext highlighter-rouge">16</code> in a <a href="https://broadinstitute.github.io/picard/explain-flags.html" rel="external nofollow noopener" target="_blank">SAM FLAG</a> indicates whether a read aligns to the reverse strand of the genome (bit 16 set iff <code class="language-plaintext highlighter-rouge">-</code> strand; bit 16 unset iff <code class="language-plaintext highlighter-rouge">+</code> strand). Hence, considering all the tags and flags as Boolean values, <code class="language-plaintext highlighter-rouge">tx</code> is negative iff bit 16 and <code class="language-plaintext highlighter-rouge">ts</code> flag are the same; <code class="language-plaintext highlighter-rouge">tx</code> is positive iff bit 16 and <code class="language-plaintext highlighter-rouge">ts</code> flag are different.</p> </div> </article> <div id="giscus_thread" style="max-width: 930px; margin: 0 auto;"> <br> <script>
      let giscusTheme = determineComputedTheme();
      let giscusAttributes = {
        src: 'https://giscus.app/client.js',
        'data-repo': 'x-zang/x-zang.github.io',
        'data-repo-id': 'R_kgDOPOeaaQ',
        'data-category': 'Announcements',
        'data-category-id': 'DIC_kwDOPOeaac4CtJVF',
        'data-mapping': 'title',
        'data-strict': '1',
        'data-reactions-enabled': '1',
        'data-emit-metadata': '0',
        'data-input-position': 'bottom',
        'data-theme': giscusTheme,
        'data-lang': 'en',
        crossorigin: 'anonymous',
        async: '',
      };

      let giscusScript = document.createElement('script');
      Object.entries(giscusAttributes).forEach(([key, value]) => giscusScript.setAttribute(key, value));
      document.getElementById('giscus_thread').appendChild(giscusScript);
    </script> <noscript>Please enable JavaScript to view the <a href="http://giscus.app/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by giscus.</a> </noscript> </div> </div> </div> <footer class="fixed-bottom" role="contentinfo"> <div class="container mt-0"> © Copyright 2025 Xiaofei Carl Zang. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script src="/assets/js/tooltips-setup.js?53023e960fbc64cccb90d32e9363de2b"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-79DTGJE6KQ"></script> <script>
    window.dataLayer = window.dataLayer || [];
    function gtag() {
      dataLayer.push(arguments);
    }
    gtag('js', new Date());

    gtag('config', 'G-79DTGJE6KQ');
  </script> <script defer src="/assets/js/google-analytics-setup.js"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>